### PR TITLE
Lichess provider

### PR DESCRIPTION
**Pull request by sukalov**
I've needed auth via Lichess, and decided to try implementing custom provider with lucia v3(beta). It worked out really well, and I'd love to share the provider code with the community. If you find it interesting enough, I'll add according documentation later this week

### CODE REVIEW

File: src/providers/lichess.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	private options: {
+		redirectURI: string;
+		scope: string[];
+	};
+
+	constructor(
+		clientId: string,
+		clientSecret: string,
+		options: {
+			redirectURI: string;
```
Comment (pilcrowonpaper) [inline src/providers/lichess.ts]: If the redirect URI is required

```ts
constructor(clientId: string, clientSecret: string, redirectURI: string) {}
```

If not,

```ts
constructor(clientId: string, clientSecret: string, options?: {
  redirectURI?: string
}) {}
```

File: src/providers/lichess.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+			scope: string[];
```
Comment (pilcrowonpaper) [inline src/providers/lichess.ts]: I think this is a mistake?
```suggestion
```

File: src/providers/lichess.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		this.clientSecret = clientSecret;
+		this.options = options;
+		this.clientId = clientId;
+	}
+
+	public async createAuthorizationURL(
+		state: string,
+		codeVerifier: string,
+		options?: {
+			scope: string[];
```
Comment (pilcrowonpaper) [inline src/providers/lichess.ts]:
```suggestion
			scopes?: string[];
```

File: src/providers/lichess.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	private clientSecret: string;
+	private clientId: string;
+	private options: {
+		redirectURI: string;
+		scope: string[];
+	};
+
+	constructor(
+		clientId: string,
+		clientSecret: string,
```
Comment (pilcrowonpaper) [inline src/providers/lichess.ts]: Does Lichess require a client secret?
https://lichess.org/api#tag/OAuth/operation/apiToken

File: src/providers/lichess.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	): Promise<URL> {
+		return await this.client.createAuthorizationURL({
+			state,
+			scopes: options?.scope ?? [],
+			codeVerifier
+		});
+	}
+
+	public async validateAuthorizationCode(code: string, codeVerifier: string): Promise<LichessTokens> {
+		const options: ValidateAuthorizationCodeRequestOptions = {
```
Comment (pilcrowonpaper) [inline src/providers/lichess.ts]: You can just pass the options to `this.client.validateAuthorizationCode()` directly to be consistent with the other providers (and remove `ValidateAuthorizationCodeRequestOptions`)

**Review by pilcrowonpaper — CHANGES_REQUESTED** [review]

File: src/providers/lichess.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	private clientId: string;
+    private redirectURI: string;
+	private options?: {
+		scopes?: string[];
+	};
+
+	constructor(
+		clientId: string,
+        redirectURI: string,
+		options?: {
```
Comment (pilcrowonpaper) [inline src/providers/lichess.ts]: This can be removed since `scopes` isn't required

**Review by pilcrowonpaper — CHANGES_REQUESTED** [review]
Can you also format it by running `pnpm format`?

**Review by pilcrowonpaper — APPROVED** [review]
Can you update the docs and the README?

**Comment by sukalov** [comment]
sure! i will

**Comment by sukalov** [comment]
HI @pilcrowOnPaper  
I've made models of Lichess api responses but couldn't find, if I can provide those somewhere in the codebase.. Is there a place for such type specifications?

**Comment by pilcrowonpaper** [comment]
Can you update `sidebar` in `docs/malta.config.json`?

**Comment by pilcrowonpaper** [comment]
We don't export the user type anymore btw

**Review by pilcrowonpaper — APPROVED** [review]

**Comment by pilcrowonpaper** [comment]
Thanks!

**Comment by sukalov** [comment]
> We don't export the user type anymore btw

yep, i've noticed it's fine
thank you! appreciate your work

**Pull request merged**